In [1]:
import pandas as pd

In [2]:
df_fusionné = pd.read_csv('df_fusionné.csv')

In [3]:
df_fusionné_binaire = df_fusionné.sample(frac=0.5)
df_fusionné_binaire.iloc[:, 0] = df_fusionné_binaire.iloc[:, 0].apply(lambda x: 0 if x == 'N' else 1)

In [4]:
# Count the occurrences of each class
counts = df_fusionné_binaire.iloc[:, 0].value_counts()

# Find the number of instances in the minority class
min_count = counts.min()

# Create balanced DataFrame through undersampling
df_balanced_under = pd.concat([
    df_fusionné_binaire[df_fusionné_binaire.iloc[:, 0] == 0].sample(min_count),
    df_fusionné_binaire[df_fusionné_binaire.iloc[:, 0] == 1].sample(min_count)
])

# Shuffle the DataFrame to mix the classes
df_balanced_under = df_balanced_under.sample(frac=1)


In [5]:
# Count occurrences of 0 and 1 in the first column
value_counts = df_balanced_under.iloc[:, 0].value_counts()

# Display the counts
print(value_counts)

ColumnName
0    18658
1    18658
Name: count, dtype: int64


In [6]:
df_balanced_under

ColumnName         0         1         2         3         4         5  \
69334           0  0.405136  0.405064  0.405044  0.405067  0.405066  0.405079   
75717           0  0.426701  0.458208  0.487934  0.524156  0.544022  0.556215   
115229          0  0.451954  0.452143  0.452455  0.452912  0.453267  0.453613   
17256           1  0.470496  0.469965  0.469848  0.470294  0.469928  0.469355   
74880           0  0.424751  0.422709  0.421053  0.419570  0.417531  0.415560   
...           ...       ...       ...       ...       ...       ...       ...   
100158          1  0.487114  0.487483  0.487889  0.488319  0.488716  0.489115   
17763           1  0.573928  0.580920  0.578595  0.560030  0.562307  0.568119   
177246          1  0.485755  0.485483  0.485129  0.484718  0.484377  0.484032   
62622           0  0.564796  0.600000  0.637333  0.647625  0.638043  0.610589   
49183           0  0.435293  0.435211  0.435099  0.434980  0.434913  0.434851   

               6         7         8  ...       710       711       712  \
69334   0.405116  0.405177  0.405290  ...  0.408419  0.408097  0.407769   
75717   0.545518  0.524765  0.512280  ...  0.434329  0.435728  0.438044   
115229  0.454060  0.454517  0.454824  ...  0.463496  0.456651  0.455129   
17256   0.469148  0.468944  0.468111  ...  0.495825  0.494378  0.495180   
74880   0.413543  0.411838  0.411806  ...  0.445907  0.446766  0.445567   
...          ...       ...       ...  ...       ...       ...       ...   
100158  0.489518  0.489932  0.490410  ...  0.485096  0.485325  0.485501   
17763   0.567412  0.565393  0.566602  ...  0.518779  0.525368  0.517737   
177246  0.483670  0.483286  0.482792  ...  0.484522  0.484551  0.484686   
62622   0.557717  0.495524  0.429442  ...  0.443885  0.450686  0.447246   
49183   0.434802  0.434742  0.434558  ...  0.435557  0.435510  0.435496   

             713       714       715       716       717       718       719  
69334   0.407429  0.407109  0.406789  0.406413  0.406042  0.405653  0.405299  
75717   0.440174  0.445251  0.450506  0.447307  0.440880  0.439213  0.448739  
115229  0.455370  0.455993  0.456928  0.456099  0.454803  0.453467  0.452193  
17256   0.496141  0.498450  0.500244  0.495731  0.489180  0.481193  0.473110  
74880   0.443443  0.441015  0.438244  0.435561  0.432882  0.429941  0.427041  
...          ...       ...       ...       ...       ...       ...       ...  
100158  0.485662  0.485840  0.486024  0.486202  0.486395  0.486583  0.486794  
17763   0.505912  0.496244  0.489036  0.485787  0.489852  0.500500  0.504228  
177246  0.484865  0.485012  0.485158  0.485356  0.485534  0.485740  0.485894  
62622   0.440505  0.443776  0.450915  0.444635  0.444866  0.448424  0.443464  
49183   0.435492  0.435483  0.435473  0.435453  0.435424  0.435397  0.435362  

[37316 rows x 721 columns]

In [7]:
import pandas as pd
from sklearn.model_selection import train_test_split

X = df_balanced_under.iloc[:, 1:]  
y = df_balanced_under.iloc[:, 0] 

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [8]:
print(y_train.unique())

[0 1]


In [9]:
y_train = y_train.astype(int)
y_test = y_test.astype(int)

In [10]:
import tensorflow as tf
from keras.models import Sequential
from keras.layers import Conv1D, BatchNormalization, MaxPooling1D, Dropout, Flatten, Dense
from keras.optimizers import Adam

Model = Sequential()

Model.add(Conv1D(512, 10, strides=1, padding="same", activation="relu", input_shape=(720, 1)))
Model.add(BatchNormalization())
Model.add(MaxPooling1D(3, strides=2, padding="same"))

Model.add(Conv1D(256, 10, strides=1, padding="same", activation="relu"))
Model.add(Dropout(0.2))
Model.add(MaxPooling1D(3, strides=2, padding="same"))

Model.add(Conv1D(128, 10, strides=1, padding="same", activation="relu"))
Model.add(Dropout(0.2))
Model.add(MaxPooling1D(3, strides=2, padding="same"))

Model.add(Conv1D(64, 10, strides=1, padding="same", activation="relu"))
Model.add(Dropout(0.2))
Model.add(MaxPooling1D(3, strides=2, padding="same"))

Model.add(Conv1D(32, 10, strides=1, padding="same", activation="relu"))
Model.add(Dropout(0.2))
Model.add(MaxPooling1D(3, strides=2, padding="same"))

Model.add(Flatten())
Model.add(Dense(1024, activation='relu'))
Model.add(Dropout(0.2))

Model.add(Dense(512, activation='relu'))
Model.add(Dropout(0.2))

Model.add(Dense(64, activation='relu'))
Model.add(Dropout(0.2))

Model.add(Dense(units=1, activation='sigmoid'))  # Use 'sigmoid' for binary classification


/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/keras/src/layers/convolutional/base_conv.py:99: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(
2024-03-13 20:01:53.151841: I metal_plugin/src/device/metal_device.cc:1154] Metal device set to: Apple M3 Pro
2024-03-13 20:01:53.151863: I metal_plugin/src/device/metal_device.cc:296] systemMemory: 18.00 GB
2024-03-13 20:01:53.151867: I metal_plugin/src/device/metal_device.cc:313] maxCacheSize: 6.00 GB
2024-03-13 20:01:53.151895: I tensorflow/core/common_runtime/pluggable_device/pluggable_device_factory.cc:305] Could not identify NUMA node of platform GPU ID 0, defaulting to 0. Your kernel may not have been built with NUMA support.
2024-03-13 20:01:53.151906: I tensorflow/core/common_runtime/pluggable_device/pluggable_device_factory.cc:271] Created TensorFlow

In [11]:
print(Model.summary())

Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ conv1d (Conv1D)                 │ (None, 720, 512)       │         5,632 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ batch_normalization             │ (None, 720, 512)       │         2,048 │
│ (BatchNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling1d (MaxPooling1D)    │ (None, 360, 512)       │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv1d_1 (Conv1D)               │ (None, 360, 256)       │     1,310,976 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout (Dropout)               │ (None, 360, 256)       │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling1d_1 (MaxPooling1D)  │ (None, 180, 256)       │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv1d_2 (Conv1D)               │ (None, 180, 128)       │       327,808 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_1 (Dropout)             │ (None, 180, 128)       │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling1d_2 (MaxPooling1D)  │ (None, 90, 128)        │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv1d_3 (Conv1D)               │ (None, 90, 64)         │        81,984 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_2 (Dropout)             │ (None, 90, 64)         │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling1d_3 (MaxPooling1D)  │ (None, 45, 64)         │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv1d_4 (Conv1D)               │ (None, 45, 32)         │        20,512 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_3 (Dropout)             │ (None, 45, 32)         │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling1d_4 (MaxPooling1D)  │ (None, 23, 32)         │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ flatten (Flatten)               │ (None, 736)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense (Dense)                   │ (None, 1024)           │       754,688 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_4 (Dropout)             │ (None, 1024)           │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 512)            │       524,800 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_5 (Dropout)             │ (None, 512)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_2 (Dense)                 │ (None, 64)             │        32,832 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_6 (Dropout)             │ (None, 64)             │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_3 (Dense)                 │ (None, 1)              │            65 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 3,061,345 (11.68 MB)

 Trainable params: 3,060,321 (11.67 MB)

 Non-trainable params: 1,024 (4.00 KB)

None


In [12]:
# Define callbacks
Early_Stopper = tf.keras.callbacks.EarlyStopping(monitor="loss", patience=3, mode="min")


In [13]:
# Compile the model
Model.compile(optimizer=Adam(), loss='binary_crossentropy', metrics=['accuracy'])

In [14]:
Conv1D_Model = Model.fit(X_train, y_train, epochs=70, validation_data=(X_test, y_test), callbacks=[Early_Stopper])

Epoch 1/70


2024-03-13 20:01:54.098815: I tensorflow/core/grappler/optimizers/custom_graph_optimizer_registry.cc:117] Plugin optimizer for device_type GPU is enabled.


933/933 ━━━━━━━━━━━━━━━━━━━━ 107s 112ms/step - accuracy: 0.7182 - loss: 0.5485 - val_accuracy: 0.6235 - val_loss: 0.6302
Epoch 2/70
933/933 ━━━━━━━━━━━━━━━━━━━━ 106s 114ms/step - accuracy: 0.8440 - loss: 0.3730 - val_accuracy: 0.6699 - val_loss: 0.6447
Epoch 3/70
933/933 ━━━━━━━━━━━━━━━━━━━━ 105s 113ms/step - accuracy: 0.8657 - loss: 0.3210 - val_accuracy: 0.8505 - val_loss: 0.3545
Epoch 4/70
933/933 ━━━━━━━━━━━━━━━━━━━━ 105s 112ms/step - accuracy: 0.8741 - loss: 0.3107 - val_accuracy: 0.8371 - val_loss: 0.3819
Epoch 5/70
933/933 ━━━━━━━━━━━━━━━━━━━━ 105s 113ms/step - accuracy: 0.8871 - loss: 0.2807 - val_accuracy: 0.8633 - val_loss: 0.3429
Epoch 6/70
933/933 ━━━━━━━━━━━━━━━━━━━━ 105s 113ms/step - accuracy: 0.8863 - loss: 0.2815 - val_accuracy: 0.8560 - val_loss: 0.3631
Epoch 7/70
933/933 ━━━━━━━━━━━━━━━━━━━━ 106s 113ms/step - accuracy: 0.8900 - loss: 0.2720 - val_accuracy: 0.8459 - val_loss: 0.3824
Epoch 8/70
933/933 ━━━━━━━━━━━━━━━━━━━━ 103s 111ms/step - accuracy: 0.8910 - loss: 0.26

In [ ]:
Model.save('Conv1D_Model.h5')  # Saves the model to a HDF5 file

In [ ]:
Model_Results = Model.evaluate(X_test,y_test)
print("LOSS:  " + "%.4f" % Model_Results[0])
print("ACCURACY:  " + "%.4f" % Model_Results[1])

95/95 ━━━━━━━━━━━━━━━━━━━━ 3s 33ms/step - accuracy: 0.8389 - loss: 0.3837
LOSS:  0.3689
ACCURACY:  0.8513


In [ ]:
# Assuming your model is named 'Model'
y_pred_probs = Model.predict(X_test)
y_pred = (y_pred_probs > 0.5).astype("int32")  # Convert probabilities to binary labels


95/95 ━━━━━━━━━━━━━━━━━━━━ 3s 31ms/step


In [ ]:
from sklearn.metrics import confusion_matrix, precision_score, recall_score, f1_score, roc_auc_score, roc_curve, log_loss

# Confusion Matrix
conf_matrix = confusion_matrix(y_test, y_pred)
print("Confusion Matrix:\n", conf_matrix)

# Precision
precision = precision_score(y_test, y_pred)
print(f"Precision: {precision:.2f}")

# Recall
recall = recall_score(y_test, y_pred)
print(f"Recall: {recall:.2f}")

# F1 Score
f1 = f1_score(y_test, y_pred)
print(f"F1 Score: {f1:.2f}")

# ROC-AUC Score
roc_auc = roc_auc_score(y_test, y_pred_probs)
print(f"ROC AUC Score: {roc_auc:.2f}")

# Now, you can calculate the log loss
logloss = log_loss(y_test, y_pred_probs)
print(f"Log Loss: {logloss:.2f}")

Confusion Matrix:
 [[1338  204]
 [ 246 1238]]
Precision: 0.86
Recall: 0.83
F1 Score: 0.85
ROC AUC Score: 0.92
Log Loss: 0.37


Confusion Matrix:
 [[1456  106]
 [ 640  834]]
Precision: 0.89
Recall: 0.57
F1 Score: 0.69
ROC AUC Score: 0.86
Log Loss: 0.65

Confusion Matrix:
 [[1357  127]
 [ 325 1187]]
Precision: 0.90
Recall: 0.79
F1 Score: 0.84
ROC AUC Score: 0.95
Log Loss: 0.33

Confusion Matrix:
 [[1401  104]
 [ 947  576]]
Precision: 0.85
Recall: 0.38
F1 Score: 0.52
ROC AUC Score: 0.81
Log Loss: 0.64


Confusion Matrix:
 [[1425  117]
 [ 375 1122]]
Precision: 0.91
Recall: 0.75
F1 Score: 0.82
ROC AUC Score: 0.93
Log Loss: 0.41
